emission/analysis/modelling/trip_model/greedy_similarity_binning.py

```diff
@@ -0,0 +1,275 @@
+        if predicted_bin_id is None:
+            logging.debug(f"unable to predict bin for trip {trip}")
+            return [], 0
+        else:
+            predictions = predicted_bin_record['predictions']
+            n_features = len(predicted_bin_record['features'])
+            logging.debug(f"found cluster {predicted_bin_id} with predictions {predictions}")
+            return predictions, n_features
+
+    def to_dict(self) -> Dict:
+        return self.bins
+
+    def from_dict(self, model: Dict):
+        self.bins = model
+
+    def extract_features(self, trip: ecwc.Confirmedtrip) -> List[float]:
+        features = self.metric.extract_features(trip)
+        return features
+
+    def _assign_bins(self, trips: List[ecwc.Confirmedtrip]):
+        """
+        assigns each trip to a bin by greedy similarity search
+        [see https://github.com/e-mission/e-mission-server/blob/5b9e608154de15e32df4f70a07a5b95477e7dbf5/emission/analysis/modelling/tour_model/similarity.py#L118]
+
+        :param data: trips to assign to bins
+        :type data: List[Confirmedtrip]
+        """
+        logging.debug(f"_assign_bins called with {len(trips)} trips")
+        for trip in trips:
+            trip_features = self.extract_features(trip)
+            trip_labels = trip['data']['user_input']
+            
+            bin_id = self._find_matching_bin_id(trip_features)
+            if bin_id is not None:
+                # add to existing bin
+                logging.debug(f"adding trip to bin {bin_id} with features {trip_features}")
+                self.bins[bin_id]['features'].append(trip_features)
+                self.bins[bin_id]['labels'].append(trip_labels)
+            else:
+                # create new bin
+                new_bin_id = str(len(self.bins))
+                new_bin_record = {
+                    "features": [trip_features],
+                    "labels": [trip_labels],
+                    "predictions": []
+                }
+                logging.debug(f"creating new bin {new_bin_id} at location {trip_features}")
+                self.bins[new_bin_id] = new_bin_record
+
+    def _find_matching_bin_id(self, trip_features: List[float]) -> Optional[str]:
+        """
+        finds an existing bin where all bin features are "similar" to the incoming
+        trip features.
+
+        :param trip_features: feature row for the incoming trip
+        :return: the id of a bin if a match was found, otherwise None
+        """
+        for bin_id, bin_record in self.bins.items():
+                matches_bin = all([self.metric.similar(trip_features, bin_sample, self.sim_thresh)
+                    for bin_sample in bin_record['features']])
+                if matches_bin:
+                    return bin_id
+        return None
+
+    def _nearest_bin(self, trip: ecwc.Confirmedtrip) -> Tuple[Optional[int], Optional[Dict]]:
+        """
+        finds a bin which contains at least all matching features. the 
```

Comment:
ah, sorry, i did mistakenly read the behavior of find_bin was also "any". we just need one of these bin search methods, let me know if you would like me to clean this up.